### ISSUE TITLE
Wrong solution for a linear under-determined system

### ISSUE THREAD

**Issue by sgsokol** [sympy/sympy#20046]
Consider the following system and its solution:
```
from sympy import *
from sympy.abc import j,f

# augmented matrix
a=Matrix([
[1,  1, 1,  1, 1,  1, 1,  1,  1],
[0, -1, 0, -1, 0, -1, 0, -1, -j],
[0,  0, 0,  0, 1,  1, 1,  1,  f]
])

x=Matrix(list(linsolve(a))[0])

# check if x is a solution
a[:,:-1]*x-a[:,-1] # should be all 0 but the second term is 'tau0-tau1'
```
Do you have an idea of how it can be fixed?

**Comment by sgsokol** [comment]
I've just found that `a.col_swap(2,4)` leads to the right solution. However, it reasonable to expect that `linsolve()` find by itself the column permutation leading to a full main diagonal and then apply the reverse permutation on the solution vector.

**Comment by sylee957** [comment]
I don't think that there is any problem with the master.
Which version are you having a problem with?

**Comment by oscarbenjamin** [comment]
I think that this was a problem in `gauss_jordan_solve` that was fixed by #19885. Also `linsolve` already stopped using `gauss_jordan_solve` since #18844.

**Comment by sgsokol** [comment]
```
Python 3.7.6 (default, Jan 21 2020, 20:43:18) 
[GCC 8.3.1 20190524] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy
>>> sympy.__version__
'1.6.2'
```

**Comment by oscarbenjamin** [comment]
Can you try with sympy master?

I think that this is already fixed but the fix isn't yet released. With master I get:
```julia
In [3]: x=Matrix(list(linsolve(a))[0])                                                                                                                        

In [4]: x                                                                                                                                                     
Out[4]: 
⎡-f - j - τ₀ + τ₂ + τ₄ + 1⎤
⎢                         ⎥
⎢    j - τ₁ - τ₂ - τ₄     ⎥
⎢                         ⎥
⎢           τ₀            ⎥
⎢                         ⎥
⎢           τ₁            ⎥
⎢                         ⎥
⎢    f - τ₂ - τ₃ - τ₄     ⎥
⎢                         ⎥
⎢           τ₂            ⎥
⎢                         ⎥
⎢           τ₃            ⎥
⎢                         ⎥
⎣           τ₄            ⎦

In [5]: a[:,:-1]*x-a[:,-1]                                                                                                                                    
Out[5]: 
⎡0⎤
⎢ ⎥
⎢0⎥
⎢ ⎥
⎣0⎦
```

**Comment by sgsokol** [comment]
By the way, is it planned to be released soon?

**Comment by oscarbenjamin** [comment]
I guess there will be a release at some point. The GSOC projects are only just finished for now and there's probably some tidying up needed after those.

**Comment by oscarbenjamin** [comment]
This issue can be closed if a test is added for this case along with the one already added in #19885.

**Comment by Ayush-Malik** [comment]
@oscarbenjamin @sylee957 Can you please tell me in which file and at which location i have to add test case for this 😀

**Comment by oscarbenjamin** [comment]
Take a look at the changes made in the other PR (#19885)

**Comment by Ayush-Malik** [comment]
```
def test_issue_20046():
    # augmented matrix
    A = Matrix([
    [1,  1, 1,  1, 1,  1, 1,  1,  1],
    [0, -1, 0, -1, 0, -1, 0, -1, -j],
    [0,  0, 0,  0, 1,  1, 1,  1,  f]
    ])

    x=Matrix(list(linsolve(A))[0])

    # check if x is A solution
    ans = A[ : , : -1 ] * x - A[ : , -1 ]
    
    actual_ans = Matrix([[0] , [0] , [0]])

    assert ans == actual_ans
```
@oscarbenjamin Is this fine , Can I create a new function named as (test_issue_20046) in ```sympy/matrices/tests/test_solvers.py ```

**Comment by oscarbenjamin** [comment]
I would just add it to the same function as was added in the other PR but change the function name to be something meaningful and use comments to link to the github issue:
```
def test_linsolve_underdetermined():
    # https://github.com/sympy/sympy/issues/20046
    example 1

    # https://github.com/sympy/sympy/issues/12345
    example 2
```
I generally dislike using an issue number as a test name as it just means that after a while all our tests are a big mess of random issue numbers.

The test (for both cases) should test both `linsolve` and `gauss_jordan_solve` ideally.

It's good to test that what is returned is a solution but that's not enough because we need to know that it is the general solution for the system so we also need to test what it is that is actually returned.

**Comment by Ayush-Malik** [comment]
```
def test_linsolve_underdetermined():
    
    # https://github.com/sympy/sympy/issues/20046
    A = Matrix([
    [1,  1, 1,  1, 1,  1, 1,  1,  1],
    [0, -1, 0, -1, 0, -1, 0, -1, -j],
    [0,  0, 0,  0, 1,  1, 1,  1,  f]
    ])


    sol_1=Matrix(list(linsolve(A))[0])

    w = {}
    for s in sol_1.atoms(Symbol):
        # Extract dummy symbols used in the solution.
        w[s.name] = s
    
    assert sol_1 == Matrix([[-f - j - w['tau0'] + w['tau2'] + w['tau4'] + 1],
                          [j - w['tau1'] - w['tau2'] - w['tau4']],
                          [w['tau0']],
                          [w['tau1']],
                          [f - w['tau2'] - w['tau3'] - w['tau4']],
                          [w['tau2']], [w['tau3']], [w['tau4']]])

    # https://github.com/sympy/sympy/issues/12345
    sol_2 = A[ : , : -1 ] * sol_1 - A[ : , -1 ]
    assert sol_2 == Matrix([[0] , [0] , [0]])
```
Is it good to go for now

**Comment by oscarbenjamin** [comment]
Yes, something like that.

Note that `https://github.com/sympy/sympy/issues/12345` is just a made up issue. The other issue is https://github.com/sympy/sympy/pull/19885 and is the one that already has a test. I meant that the tests for both issues should be in a single function.

Also the symbols can be created with
```
t0, t1, t2, t3, t4 = symbols('t:5')
```
rather than using `.atoms(Symbol)`.

**Comment by Ayush-Malik** [comment]
@oscarbenjamin ok I am on it

**Comment by Ayush-Malik** [comment]
```
from sympy import *
from sympy.abc import j,f
def test_linsolve_underdetermined():
    # https://github.com/sympy/sympy/issues/20046
    A = Matrix([
    [1,  1, 1,  1, 1,  1, 1,  1,  1],
    [0, -1, 0, -1, 0, -1, 0, -1, -j],
    [0,  0, 0,  0, 1,  1, 1,  1,  f]
    ])

    sol_1=Matrix(list(linsolve(A))[0])

    tau0, tau1, tau2, tau3, tau4 = symbols('tau:5')

    assert sol_1 == Matrix([[-f - j - tau0 + tau2 + tau4 + 1],
                          [j - tau1 - tau2 - tau4],
                          [tau0],
                          [tau1],
                          [f - tau2 - tau3 - tau4],
                          [tau2],
                          [tau3],
                          [tau4]])

    # https://github.com/sympy/sympy/issues/19885
    sol_2 = A[ : , : -1 ] * sol_1 - A[ : , -1 ]
    assert sol_2 == Matrix([[0] , [0] , [0]])
test_linsolve_underdetermined()
```
@oscarbenjamin , There is a test case already in ``` sympy/matrices/tests/test_solvers.py ```  (test_issue_19815) associated with issue #19885 in which gauss_jordan solve is checked do i have to add something to it

**Comment by oscarbenjamin** [comment]
Yes, this case can be added to that function and it should be renamed.

**Comment by oscarbenjamin** [comment]
Fixed in #20074
